editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,627 @@
+				rect = rect.grow(-handle_thickness);
+				border_col.a = 0.54;
+				draw_rect(rect, border_col, false, handle_thickness);
+			}
+		}
+	}
+
+	// Draw "button" for color selector.
+	int button_offset = total_w + handle_width + draw_spacing;
+	if (selected_index != -1) {
+		Color grabbed_col = gradient->get_color(selected_index);
+		if (grabbed_col.a < 1) {
+			draw_texture_rect(get_editor_theme_icon(SNAME("GuiMiniCheckerboard")), Rect2(button_offset, 0, h, h), true);
+		}
+		draw_rect(Rect2(button_offset, 0, h, h), grabbed_col);
+		if (grabbed_col.r > 1 || grabbed_col.g > 1 || grabbed_col.b > 1) {
+			// Draw an indicator to denote that the currently selected color is "overbright".
+			draw_texture(get_theme_icon(SNAME("overbright_indicator"), SNAME("ColorPicker")), Point2(button_offset, 0));
+		}
+	} else {
+		// If no color is selected, draw grey color with 'X' on top.
+		draw_rect(Rect2(button_offset, 0, h, h), Color(0.5, 0.5, 0.5, 1));
+		draw_line(Vector2(button_offset, 0), Vector2(button_offset + h, h), Color(0.8, 0.8, 0.8));
+		draw_line(Vector2(button_offset, h), Vector2(button_offset + h, 0), Color(0.8, 0.8, 0.8));
+	}
+}
+
+void GradientEdit::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			picker->connect("color_changed", callable_mp(this, &GradientEdit::_color_changed));
+		} break;
+		case NOTIFICATION_THEME_CHANGED: {
+			draw_spacing = BASE_SPACING * get_theme_default_base_scale();
+			handle_width = BASE_HANDLE_WIDTH * get_theme_default_base_scale();
+		} break;
 		case NOTIFICATION_DRAW: {
-			Ref<Texture2D> icon = get_editor_theme_icon(SNAME("ReverseGradient"));
-			if (is_pressed()) {
-				draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()), false, get_theme_color(SNAME("icon_pressed_color"), SNAME("Button")));
-			} else {
-				draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()));
+			_redraw();
+		} break;
+		case NOTIFICATION_MOUSE_EXIT: {
+			if (hovered_index != -1) {
+				hovered_index = -1;
+				queue_redraw();
+			}
+		} break;
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (!is_visible()) {
+				grabbing = GRAB_NONE;
 			}
 		} break;
 	}
 }
 
-Size2 GradientReverseButton::get_minimum_size() const {
-	return (get_editor_theme_icon(SNAME("ReverseGradient"))->get_size() + Size2(margin * 2, margin * 2));
+void GradientEdit::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_selected_index", "index"), &GradientEdit::set_selected_index);
+}
+
+GradientEdit::GradientEdit() {
+	set_focus_mode(FOCUS_ALL);
+
+	popup = memnew(PopupPanel);
+	picker = memnew(ColorPicker);
+	popup->add_child(picker);
+	popup->connect("about_to_popup", callable_mp(EditorNode::get_singleton(), &EditorNode::setup_color_picker).bind(GradientEdit::get_picker()));
+
+	preview_texture.instantiate();
+	preview_texture->set_width(1024);
+	add_child(popup, false, INTERNAL_MODE_FRONT);
```

Comment:
Why is it here and not next to the node allocating code?

Reply:
I think I didn't change this code, I assume it was just like "Setting up everything before adding it" sorta thing?

Reply:
For style consistency it's better to move it to where we memnew the node, I think.